examples_tests/20.Megatexture/main.cpp

```diff
@@ -84,188 +71,26 @@ layout (push_constant) uniform Block {
-#else
-    uint64_t outstandingSampleMask = ballotARB(true);
-    // maybe unroll a few times manually
-    while (outstandingSampleMask!=uint64_t(0u))
-    {
-		uvec2 tmp = unpackUint2x32(outstandingSampleMask);
-        uint subgroupFormatID = readInvocationARB(formatID,tmp[1]!=0u ? 32u:findLSB(tmp[0]));
-        bool canSample = subgroupFormatID==formatID; // do I need this? && (outstandingSampleMask&gl_SubGroupEqMaskARB)==gl_SubGroupEqMaskARB;
-        outstandingSampleMask ^= ballotARB(canSample);
-        if (canSample)
-        {
-            hiMip_retval = textureGrad(physPgTex[subgroupFormatID],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-            if (haveToDoTrilinear)
-                loMip = textureGrad(physPgTex[subgroupFormatID],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-        }
-    }
-#endif
-    if (haveToDoTrilinear)
-	    hiMip_retval = mix(hiMip_retval,loMip,LoD-float(LoD_high));
-    return hiMip_retval;
-}
-
-uvec2 unpackWrapModes(in uvec2 texData)
+uint getPgTabSzLog2(in uint _formatID)
 {
-    return (texData>>uvec2(28u,30u)) & uvec2(0x03u);
+    return precomputed.pgtab_sz_log2[_formatID];
 }
-uint unpackMaxMipInVT(in uvec2 texData)
+float getPhysPgTexSzRcp(in uint _formatID)
 {
-    return (texData.y>>24)&0x0fu;
+    return precomputed.phys_pg_tex_sz_rcp[_formatID];
 }
-vec3 unpackVirtualUV(in uvec2 texData)
+float getVTexSzRcp(in uint _formatID)
 {
-	// assert that PAGE_SZ_LOG2<8 , or change the line to uvec3(texData.yy<<uvec2(PAGE_SZ_LOG2,PAGE_SZ_LOG2-8u),texData.y>>16u)
-    uvec3 unnormCoords = uvec3(texData.y<<PAGE_SZ_LOG2,texData.yy>>uvec2(8u-PAGE_SZ_LOG2,16u))&uvec3(uvec2(0xffu)<<PAGE_SZ_LOG2,0xffu);
-    return vec3(unnormCoords);
+    return precomputed.vtex_sz_rcp[_formatID];
 }
-vec2 unpackSize(in uvec2 texData)
-{
-	return vec2(texData.x&0xffffu,texData.x>>16u);
-}
-#define WRAP_REPEAT 0u
-#define WRAP_CLAMP 1u
-#define WRAP_MIRROR 2u
-float wrapTexCoord(float tc, in uint mode)
-{
-    switch (mode)
-    {
-    case WRAP_REPEAT: tc = fract(tc); break;
-    case WRAP_CLAMP:  tc = clamp(tc, 0.0, 1.0); break;
-    case WRAP_MIRROR: tc = 1.0 - abs(mod(tc,2.0)-1.0); break;
-    default: break;
-    }
-    return tc;
-}
-vec4 textureVT(in uvec2 _texData, in vec2 uv, in mat2 dUV)
-{
-    vec2 originalSz = unpackSize(_texData);
-	dUV[0] *= originalSz;
-	dUV[1] *= originalSz;
 
-    uvec2 wrap = unpackWrapModes(_texData);
-    uv.x = wrapTexCoord(uv.x,wrap.x);
-    uv.y = wrapTexCoord(uv.y,wrap.y);
-
-	vec3 virtualUV = unpackVirtualUV(_texData);
-
-    uint formatID = getPhysicalIDForLayer(uint(virtualUV.z));
-
-    virtualUV.xy += uv*originalSz;
-    virtualUV.xy *= precomputed.vtex_sz_rcp[formatID];
-
-    return vTextureGrad(formatID, virtualUV, dUV, int(unpackMaxMipInVT(_texData)));
-}
+#include <irr/builtin/glsl/virtual_texturing/functions.glsl/6/6/7/8/pageTable/physPgTex/getPgTabSzLog2/getPhysPgTexSzRcp/getVTexSzRcp/getPhysicalIDForLayer>
```

Comment:
pretty path

Reply:
need more semantics about what the `/6/6/7/8/pageTable/physPgTex/getPgTabSzLog2/getPhysPgTexSzRcp/getVTexSzRcp/getPhysicalIDForLayer` mean...